File: Rubyとか/12_EuclideanAlgorithm_CommonPrimeDivisors.rb
# you can write to stdout for debugging purposes, e.g.
# puts "this is a debug message"

# you can write to stdout for debugging purposes, e.g.
# puts "this is a debug message"


def GCD(m, n)
	return 0 if m == 0 || n == 0
	while m != 0 && n != 0
		if m > n then
			m %= n
		else
			n %= m
		end
	end
	return [m, n].max
end

def LCM(m, n)
	return 0 if m == 0 || n == 0
	return m * n / GCD(m, n)
end


def solution(a, b)
  # write your code in Ruby 2.2
  ans = 0
  a.length.times do |i|
    aa = a[i]; bb = b[i]
    lcm = LCM(aa, bb)
    gcd = GCD(aa, bb)

    # gcdで割り続けると良い感じ
    # [2,2,3] と [2,3,3]だと6で割って[2],[3]
    # → [2], [3] と 6のgcdで割ると[1], [1]

    # [2,2,3] と [2,3,3,5]だと 6でわって[2], [3,5]
    # → [2], [3, 5] と 6のgcdで割ると[1], [5]
    # → 6 と [5]のgcdで割っても5 

    # 初期のGCDと更新した空間で割り続ければ初期のGCDに含まれないやーつが出るはず
    prev = -1
    while aa != prev 
      prev = aa
      aa = aa / GCD(aa, gcd)
    end 

    prev = -1
    while bb != prev 
      prev = bb 
      bb = bb / GCD(bb, gcd)
    end 

    #puts "gcd:#{gcd} aa:#{aa} bb:#{bb}"
    ans += 1 if aa * bb == 1
  end
  return ans
end

File: Python/12_EuclideanAlgorithm_CommonPrimeDivisors.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")

def gcd(a, b):
    if a * b == 0: return 0
    while a * b != 0:
        a, b = max(a, b), min(a, b)
        a = a % b 
    return int(max(a, b))
# end of func 

def lcm(a, b):
    if a * b == 0: return 0 
    return int(a * b / gcd(a, b))
# end of lcm

def solution(A, B):
    # write your code in Python 3.6
    n = len(A)
    ans = 0 
    for i in range(n):
        a, b = A[i], B[i]
        g = gcd(a, b) 

        # a=15, b=75 のように 共通のPrimedivisorを持つ場合
        # a = 3^1 * 5^1
        # b = 3^1 * 5^2 のように指数部分が異なるだけである
        # → 3^x * 5^y (0 <= x, y)で割り続ければ a,b は1になる
        # → GCD(a, GCD(a,b))のようにするとx,yをいい感じに決められる

        # a=10, b=30 のように共通のPrimeDivisorを持たない場合
        # a = 2^1 * 5^1
        # b = 2^3 * 5^1 * 3^1 のように底が異なるので
        # GCD(b, GCD(a, b))を繰り返しても a に 3 が含まれていないのでGCDに3の倍数が出ることはない
        # → 2^x * 5^y (0 <= x, y) で割り続けると 3^z みたいなのが残る
       
        gg = g
        while gg > 1:
            gg = gcd(a, g)
            a = a / gg  
    
        gg = g
        while gg > 1:
            gg = gcd(b, g)
            b = b / gg  

        if a * b == 1: ans += 1
    # end of for

    return ans
# end of func


